### DISCUSSION TITLE
Only add WGPUTextureUsage_CopySrc when required

### DISCUSSION THREAD

**Discussion by guusw** [fragcolor-xyz/shards#1217] [Ideas]
Currently WGPUTextureUsage_CopySrc is always set for render targets.
It should not be set but added whenever usage of CopySrc is required and then recreate the texture.

**Comment by sinkingsugar** [comment]
@claude can you check if this is still an issue and propose a fix if possible? if too convoluted we might archive it as nice to have, as gfx not a high priority.

**Comment by claude** [comment]
**Claude finished @sinkingsugar's task** —— [View job](https://github.com/fragcolor-xyz/shards/actions/runs/19521862082)

---
### Investigating WGPUTextureUsage_CopySrc Issue

- [x] Gather context on current WGPUTextureUsage_CopySrc usage
- [x] Analyze if this is still an issue
- [x] Propose a fix or provide recommendation

---

### Analysis

**Current State:** The issue is still present in the codebase.

In `shards/gfx/texture.cpp:159-160`, `WGPUTextureUsage_CopySrc` is unconditionally added to all render attachment textures:

```cpp
if (textureFormatFlagsContains(desc.format.flags, TextureFormatFlags::RenderAttachment)) {
    wgpuDesc.usage |= WGPUTextureUsage_RenderAttachment | WGPUTextureUsage_CopySrc;
}
```

**Why CopySrc is needed:** This flag enables `wgpuCommandEncoderCopyTextureToBuffer()` operations for:
- Texture readback to CPU (`Renderer::copyTexture()` in `shards/gfx/renderer.cpp:411`)
- Test frame capture (`TestRenderer::getTestFrame()`)

**Impact:** Setting `CopySrc` on all render targets may prevent certain GPU driver optimizations. Some platforms/drivers can use more efficient memory layouts when they know a render target won't be read back.

---

### Recommendation: Archive as "Nice to Have"

**Implementing this properly would be quite involved:**

1. **New TextureFormatFlag** - Add `TextureFormatFlags::CopySrc` (~0x20)

2. **Runtime Detection** - When `copyTexture()` is called, check if the texture has `CopySrc` usage. If not:
   - Set the flag: `texture->initWithFlags(flags | TextureFormatFlags::CopySrc)`
   - Wait for texture recreation (version bump triggers this)
   - Proceed with copy

3. **Synchronization Issues** - Texture recreation mid-frame would invalidate:
   - Active command encoders referencing the old texture
   - Pipeline bind groups
   - Texture views in render passes

4. **Alternative Approach** - Lazy recreation before first use:
   - Track if `CopySrc` was requested
   - Recreate texture before first render pass if flags changed
   - Requires careful frame sequencing

**Why it's convoluted:**
- The texture may already be bound in an active render pass when `copyTexture()` is called
- WebGPU doesn't allow recreating textures while they're in use
- Would need to delay the recreation to the next frame, complicating the API

**Expected benefit:** Marginal performance improvement on some GPU/driver combinations, only when many render targets exist that are never read back.

---

### Verdict

Given that GFX is not a high priority and the fix would require significant architectural changes to handle texture recreation safely during the render pipeline, I recommend **archiving this as "nice to have"**. The current approach of always enabling `CopySrc` is correct and safe, just potentially suboptimal.

If this becomes a performance bottleneck in the future, a simpler approach would be to add an explicit `TextureFormatFlags::CopySrc` flag that users must set when creating textures they intend to read back, rather than implementing lazy recreation.

---
